homeassistant/components/picnic/services.py

```diff
@@ -51,15 +50,14 @@ async def async_add_product_service(call: ServiceCall):
     )
 
 
-async def get_api_client(hass: HomeAssistant, device_id: str) -> PicnicAPI:
+async def get_api_client(hass: HomeAssistant, config_entry_id: str) -> PicnicAPI:
     """Get the right Picnic API client based on the device id, else get the default one."""
-    registry = device_registry.async_get(hass)
-    if not (device := registry.async_get(device_id)):
-        raise PicnicServiceException(f"Device with id {device_id} not found!")
-
-    if config_entry_id := next(iter(device.config_entries), None):
+    try:
         return hass.data[DOMAIN][config_entry_id][CONF_API]
-    raise PicnicServiceException(f"Device with id {device_id} not found!")
+    except KeyError:
+        raise PicnicServiceException(
```

Comment:
There's a pylint warning here.

I still think its simpler to rely on an explicit check rather than a `KeyError` since the key errors can also happen for unexpected cases.

Also, I think `HomeAsssistantError` needs to be thrown from a service call so that home assistant sees it properly and stops an automation.

As a result of both of those considerations I suggest going with:

```
if config_entry_id not in hass.data[DOMAIN]:
    raise HomeAsssistantError(f"Config entry with id {config_entry_id} not found!")
return hass.data[DOMAIN][config_entry_id][CONF_API]
```

Reply:
Shouldn't it be an `IntegrationError` or `PlatformNotReady` exception btw? They are derived from `HomeAsssistantError` but seem to be more specific for a certain situation.

Reply:
I get your point about trying to find a more specific exception. Really this is about invalid user input. I had a similar discussion here https://github.com/home-assistant/core/pull/78889#discussion_r983851581 about whether it should be `ValueError` or `vol.Invalid` for basically the exact same case of a config entry not being found.

Reply:
So should `HomeAsssistantError` also be raised in `handle_add_product()` when the user inputs a wrong ID? Or is `PicnicServiceException` fine there?

Reply:
My read is it should be either `ValueError` or `vol.Invalid`